cartography/intel/gcp/permission_relationships.py

```diff
@@ -0,0 +1,395 @@
+    return False
+
+
+def evaluate_permission_for_permission(permissions: dict, permission: str) -> bool:
+    if not permissions.get("permissions"):
+        return False
+    for clause in permissions["permissions"]:
+        if evaluate_clause(clause, permission):
+            return True
+    return False
+
+
+def evaluate_policy_binding_for_permissions(
+    assignment_data: dict[str, Any],
+    permissions: list[str],
+    resource_scope: str,
+) -> bool:
+    permissions_dict = assignment_data["permissions"]
+    scope = assignment_data["scope"]
+
+    # Level 1: Check scope matching
+    if not evaluate_scope_for_resource({"scope": scope}, resource_scope):
+        return False
+
+    for permission in permissions:
+        # Level 2: Check denied permissions
+        if not evaluate_denied_permission_for_permission(permissions_dict, permission):
+            # Level 3: Check allowed permissions
+            if evaluate_permission_for_permission(permissions_dict, permission):
+                return True
+
+    return False
+
+
+def principal_allowed_on_resource(
+    policy_bindings: dict[str, Any],
+    resource_scope: str,
+    permissions: list[str],
+) -> bool:
+    if not isinstance(permissions, list):
+        raise ValueError("permissions is not a list")
+
+    for _, assignment_data in policy_bindings.items():
+        if evaluate_policy_binding_for_permissions(
+            assignment_data, permissions, resource_scope
+        ):
+            return True
+
+    return False
+
+
+def calculate_permission_relationships(
+    principals: dict[str, Any],
+    resource_dict: dict[str, str],
+    permissions: list[str],
+) -> list[dict[str, Any]]:
+    allowed_mappings: list[dict[str, Any]] = []
+    for resource_id, resource_scope in resource_dict.items():
+        for principal_email, policy_bindings in principals.items():
+            if principal_allowed_on_resource(
+                policy_bindings, resource_scope, permissions
+            ):
+                allowed_mappings.append(
+                    {
+                        "principal_email": principal_email,
+                        "resource_id": resource_id,
+                    }
+                )
+    return allowed_mappings
+
+
+@timeit
+def get_principals_for_project(
+    neo4j_session: neo4j.Session, project_id: str
+) -> dict[str, Any]:
+    """
+    Get all principals (users, service accounts, groups) with their policy bindings
+    for a given GCP project.
+    """
+    get_principals_query = """
+    MATCH
+    (project:GCPProject{id: $ProjectId})-[:RESOURCE]->
+    (binding:GCPIAMPolicyBinding)-[:GRANTS_ROLE]->
+    (role:GCPRole)
+    MATCH
+    (principal:GCPPrincipal)-[:HAS_ALLOW_POLICY]->(binding)
+    RETURN
+    DISTINCT principal.email as principal_email, binding.id as binding_id,
+    binding.resource as binding_resource, role.permissions as role_permissions
+    """
+
+    results = neo4j_session.run(get_principals_query, ProjectId=project_id)
```

Comment:
Definitely do not use `neo4j_session.run()`. Use a managed transaction function, look for other examples